[SPARK-9674] Re-enable ignored test in SQLQuerySuite

andrewor14 opened the pull request:
The original code that this test tests is removed in https://github.com/apache/spark/commit/9270bd06fd0b16892e3f37213b5bc7813ea11fdd. It was ignored shortly before that so we never caught it. This patch re-enables the test and adds the code necessary to make it pass.

@JoshRosen @yhuai

Commit:
Fix test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40123 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40123/consoleFull) for   PR 8015 at commit [`0be3a42`](https://github.com/apache/spark/commit/0be3a425509dbe40ea9ef1b9a9b02413773eec45).

Commit:
Track aggregation memory for both sort and hash

Previous commit only tracked it for hash-based aggregation.
We should track it in both cases.

Commit:
Fix NPE

Commit:
Merge branch 'master' of github.com:apache/spark into SPARK-9674

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40132/consoleFull) for   PR 8015 at commit [`225eac2`](https://github.com/apache/spark/commit/225eac229c34c1397d311da381c63b1942addee9).

SparkQA commented:
[Test build #40123 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40123/console) for   PR 8015 at commit [`0be3a42`](https://github.com/apache/spark/commit/0be3a425509dbe40ea9ef1b9a9b02413773eec45).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #40132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40132/console) for   PR 8015 at commit [`225eac2`](https://github.com/apache/spark/commit/225eac229c34c1397d311da381c63b1942addee9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40177 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40177/consoleFull) for   PR 8015 at commit [`225eac2`](https://github.com/apache/spark/commit/225eac229c34c1397d311da381c63b1942addee9).

SparkQA commented:
[Test build #40177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40177/console) for   PR 8015 at commit [`225eac2`](https://github.com/apache/spark/commit/225eac229c34c1397d311da381c63b1942addee9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

yhuai commented:
`TungstenAggregationIterator` part looks good.

yhuai commented:
LGTM

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala:
```diff
@@ -397,14 +397,20 @@ class TungstenAggregationIterator(
   ///////////////////////////////////////////////////////////////////////////
-  // Part 4: The function used to switch this iterator from hash-based
-  // aggregation to sort-based aggregation.
+  // Part 3: Methods and fields used by sort-based aggregation.
```
oops this should be Part 4

yhuai replied:
I will have a PR to change other parts of this class. I can update the comment.

yhuai commented:
Merging to master and branch 1.5.

andrewor14 commented:
The right JIRA number is actually SPARK-9677 by the way!
